scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CardinalityGoal.java

```diff
@@ -210,6 +244,30 @@ public Collection<Conflict> getConflicts(Plan plan, final SimulationResults simu
     return conflicts;
   }
 
+  private boolean stuckInsertingZeroDurationActivities(final Plan plan){
+    if(this.durationRange != null){
+      final var inserted = plan.getEvaluation().forGoal(this).getInsertedActivities();
+      for(var insertedActivity : inserted){
+        //new activity inserted to satisfy this goal
+        if(!insertedSoFar.containsKey(insertedActivity.getId())){
+          //record it
+          insertedSoFar.put(insertedActivity.getId(), insertedActivity);
+          //if duration is zero, contribute to current no progress counter
+          if(insertedActivity.getDuration().isZero()) {
+            this.stepsWithoutProgress++;
```

Comment:
I'm not sure I agree with this definition of progress - I think it's okay if we added one zero duration activity, as long as we also added another activity with a non-zero duration, right?

I would think that progress is if the total duration has increased since the last round. (and lack of progress is if it has not)

Also - if we are trying to reach a certain number of occurrences, I think it's okay to place zero-duration activities. It's only if we have one of these cases that placing zero duration activities becomes a problem:
1. We have already satisfied the number of occurrences
2. Or we only had a durationRange goal, not an occurrences goal

Reply:
I have rewritten this to follow your idea.